File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.js
class MinHeap {
    constructor(compare) {
        this.heap = [];
        this.compare = compare;
    }

    size() {
        return this.heap.length;
    }

    peek() {
        return this.heap[0];
    }

    push(value) {
        this.heap.push(value);
        let idx = this.heap.length - 1;

        while (idx > 0) {
            const parent = Math.floor((idx - 1) / 2);
            if (this.compare(this.heap[parent], this.heap[idx]) <= 0) break;
            [this.heap[parent], this.heap[idx]] = [this.heap[idx], this.heap[parent]];
            idx = parent;
        }
    }

    pop() {
        if (this.heap.length === 1) return this.heap.pop();
        const top = this.heap[0];
        this.heap[0] = this.heap.pop();

        let idx = 0;
        while (true) {
            let left = idx * 2 + 1;
            let right = idx * 2 + 2;
            let smallest = idx;

            if (left < this.heap.length && this.compare(this.heap[left], this.heap[smallest]) < 0) smallest = left;
            if (right < this.heap.length && this.compare(this.heap[right], this.heap[smallest]) < 0) smallest = right;

            if (smallest === idx) break;
            [this.heap[idx], this.heap[smallest]] = [this.heap[smallest], this.heap[idx]];
            idx = smallest;
        }

        return top;
    }
}

function solution(jobs) {
    jobs.sort((a, b) => a[0] - b[0] || a[1] - b[1]);

    const pq = new MinHeap((a, b) => a[1] - b[1] || a[0] - b[0] || a[2] - b[2]);

    let time = 0;
    let idx = 0;
    let total = 0;
    let jobNo = 0;

    while (idx < jobs.length || pq.size() > 0) {
        while (idx < jobs.length && jobs[idx][0] <= time) {
            pq.push([jobs[idx][0], jobs[idx][1], jobNo++]);
            idx++;
        }

        if (pq.size() === 0) {
            time = jobs[idx][0];
            continue;
        }

        const [request, duration] = pq.pop();
        time += duration;
        total += time - request;
    }

    return Math.floor(total / jobs.length);
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        Arrays.sort(jobs, (a, b) -> {
            if (a[0] == b[0]) return a[1] - b[1];
            return a[0] - b[0];
        });

        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> {
            if (a[1] == b[1]) {
                if (a[0] == b[0]) return a[2] - b[2];
                return a[0] - b[0];
            }
            return a[1] - b[1];
        });

        int time = 0;
        int idx = 0;
        int total = 0;
        int n = jobs.length;
        int jobNo = 0;

        while (idx < n || !pq.isEmpty()) {
            while (idx < n && jobs[idx][0] <= time) {
                pq.offer(new int[]{jobs[idx][0], jobs[idx][1], jobNo++});
                idx++;
            }

            if (pq.isEmpty()) {
                time = jobs[idx][0];
                continue;
            }

            int[] cur = pq.poll();
            time += cur[1];
            total += time - cur[0];
        }

        return total / n;
    }
}
